Backlog calculation improvements

mmaslankaprv opened the pull request:
The backlog calculation was not correct as the translation from kafka to
log offset was incorrectly made. Fixed the backlog calculation logic to
return '0' when all batches are translated.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
dl/translation: fixed backlog calculation

The backlog calculation was not correct as the translation from kafka to
log offset was incorrectly made. Fixed the backlog calculation logic to
return '0' when all batches are translated.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
tests: added test validating translation backlog calculation

Signed-off-by: Michał Maślanka <michal@redpanda.com>

nvartolomei commented on tests/rptest/tests/datalake/throttling_test.py:
```diff
@@ -116,3 +116,44 @@ def test_basic_throttling(self, cloud_storage_type, catalog_type):
+            # restart all nodes
+            self.redpanda.restart_nodes(self.redpanda.nodes)
+            # check backlog again
+            wait_until(lambda: collect_backlog_metric() == 0,
```
Should/Can this be a stricter check? Must be 0? Maybe across couple of checks? Why there is a wait if it was previously already 0?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5e2be688-b675-4df4-8c9e-263628689352-->
<!--TEST RESULTS-->
<!--97011681-9abe-4c97-873f-c958ff9f4870-->
<!--TEST RESULTS-->
<!--cd2c4e07-fcb2-4e0e-8353-36dbe678fc28-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63866">build#63866</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63866#0195dd24-a238-4191-820e-7c1e26c4912d | FLAKY | 1/2 |
| rptest.tests.idempotency_test.IdempotencySnapshotDelivery.test_recovery_after_snapshot_is_delivered | ducktape | https://buildkite.com/redpanda/redpanda/builds/63866#0195dd40-1396-46d6-a460-aad14126a825 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63866#0195dd40-1396-46d6-a460-aad14126a825 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade_with_rollback.upgrade_after_rollback=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63866#0195dd24-a23b-42f6-8ace-6d92f107b3d1 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63973">build#63973</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63973#0195ec1e-e559-493e-bd73-35bc70b2f20f | FLAKY | 1/4 |
| rptest.tests.rack_aware_replica_placement_test.RackAwarePlacementTest.test_rack_awareness_after_node_operations.rack_layout=ABCDEF | ducktape | https://buildkite.com/redpanda/redpanda/builds/63973#0195ec1e-e558-4ef8-b18a-3c5deff0a372 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63973#0195ec23-eb2b-4e98-97eb-bd9268c683d6 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64015">build#64015</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64015#0195f04f-b0ee-44b1-a191-5e5c99d79c7e | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64015#0195f04f-b0ec-4ac0-a7b4-a6c86c23b8d8 | FLAKY         | 1/2      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64015#0195f054-0e1b-42e2-a90e-b679b84df004 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64015#0195f054-0e1c-4cee-ab0a-cbffc3636398 | FLAKY         | 1/3      |
</details>
<!--/body-->

mmaslankaprv committed:
dl/translation: track inflight translation offset in lag tracker

When calculating translation backlog we should use an inflight
translation offset to make sure backlog calculation is accurate and
includes current inflight translation, otherwise the backlog would not
take the currently translated parquet file which by default is 32MiB in
size.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

andrwng commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -103,7 +103,7 @@ class translation_stm final : public raft::persisted_stm<> {
     void update_highest_translated_offset(kafka::offset new_offset);
 
-    kafka::offset _highest_translated_offset{};
+    std::optional<kafka::offset> _highest_translated_offset;
```
Can you add a comment describing the distinction between nullopt and kafka::offset{}? Or is this never kafka::offset{}?

mmaslankaprv replied:
I simplified this as it doesn't makes sense to have optional and not initialised as they have the same semantics

andrwng reviewed (dismissed):
High level makes sense to me

vbotbuildovich commented:
/backport v25.1.x
